### DISCUSSION TITLE
MIDI Host pico2040

### DISCUSSION THREAD

**Discussion by artouste** [hathach/tinyusb#1218] [Q&A]
Hello

I discovered the raspberry pico2040 with the portage under arduino IDE from https://github.com/earlephilhower/arduino-pico

I would have liked to know in the current state if it is possible to use the pico2040 as a USB Host with a USB MIDI keyboard.

If so, please give me a starting point to use it like this.

Thank you for your contribution

**Comment by rppicomidi** [comment]
@artouste No, not yet. I am working on it. See the details in Pull Request #1219

**Comment by rppicomidi** [comment]
@hathach My MIDI host pull request #1219 is buggy and hopelessly out of date. It contains a bulk transfer implementation for RP2040 that is not as clean as the one in TinyUSB. It also contains changes outside of the MIDI driver code that are no longer required because TinyUSB implemented them in a different way. Pull request #1627 is based on my code but is much better. It fixes bugs in my code and it uses the TinyUSB RP2040 bulk transfer code. It also brought in some changes that are not really needed to implement USB MIDI host. So if you want to review an existing pull request, I would review #1627. Meanwhile, I believe I will publish on GitHub an application level MIDI Host driver. You might find that code simpler to merge to the stack.

**Reply by hathach** [reply]
thank you @rppicomidi for the sum up, it is very helpful, I will do that when I review PR

**Comment by rppicomidi** [comment]
You are welcome, @hathach. You can find the promised application MIDI Host driver code [here](https://github.com/rppicomidi/usb_midi_host).
